[RFC] drivers: pwm: API naming changes and pwm_dt_spec

gmarull opened the pull request:
This RFC attempts to improve the PWM API in multiple aspects, mainly:

1. All PWM API functions take a `pwm` parameter that indicates the selected PWM channel. The variable name (pwm) and its documentation “PWM pin” is misleading. This patch changes it to “channel” in line with the terminology used in Devicetree (see `DT_PWMS_CHANNEL...` family of macros). Most drivers refer to `pwm` as “channels”, suggesting the new name makes sense.
2. The PWM API operates on “channels”, not “pins”. While the API calls could have been changed by `_channel`, this PR takes the approach of just dropping `_pin`, e.g., `pwm_pin_set_cycles()` -> `pwm_set_cycles()`. The main reason for this choice is that all API calls operate by definition on a channel basis.
3. Similar to other APIs, this patch introduces the `pwm_dt_spec` structure and associated helpers, e.g., `PWM_DT_SPEC_GET()` or `pwm_set_cycles_dt()`. The `pwm_dt_spec` reduces the boilerplate code needed when using Devicetree. For example, if we have:

```
    led: led {
        ...
        pwms = <&pwm0 1 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
        ...
    };
```
One can do now in the application:

```c
    struct pwm_dt_spec led = PWM_DT_SPEC_GET(DT_NODELABEL(led));
    pwm_set_pulse_dt(&led, PWM_MSEC(10));
```
4. Some samples/drivers have been updated to show usage of `pwm_dt_spec`.
5. PWM `set` API operates in nanoseconds, other units/scales can be specified using e.g. `PWM_MSEC()` macros that convert down to nanoseconds.

gmarull committed:
drivers: pwm: s/pwm/channel

All PWM API functions take a "pwm" parameter that indicates the selected
PWM channel. The variable name (pwm) and its documentation "PWM pin" is
misleading. This patch changes it to "channel" in line with the
terminology used in Devicetree (see DT_PWMS_CHANNEL... family of
macros).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: remove _pin from API calls

In a first place, the PWM API operates on "channels", not "pins". While
the API calls could have been changed by _channel, this patch takes the
approach of just dropping _pin. The main reason is that all API calls
operate by definition on a channel basis, so it is a bit redundant to
make this part of the name. Because the `_dt` variants of the calls are
going to be introduced soon, the change to `_channels` + `_dt` would
make API function names quite long.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: use new API naming (no pin)

Use the API names without `_pin`.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers/samples/tests: remove usage of deprecated PWM APIs

Use the new API calls that remove pin naming.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: make all drivers use channel variable name

The variable indicating the PWM channel is now names "channel" instead
of "pwm", adjust all drivers.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: ite_it8xxx2: remove wrong ARG_UNUSED

The pwm arg is used in the function, so placing ARG_UNUSED is wrong.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: ite_it8xxx2: add missing braces

Multiple if/else blocks had missing braces, add them as this violates
Zephyr coding guidelines.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: gecko: fix access to timer register

The timer registers are accessible via the device config field, driver
code was wrong in one case (pwm is the variable indicating PWM channel).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: add pwm_dt_spec and associated helpers

Similar to other APIs, this patch introduces the pwm_dt_spec structure
and associated helpers, e.g. PWM_DT_SPEC_GET() or pwm_set_cycles_dt().

The pwm_dt_spec reduces the boilerplate code needed when using
Devicetree. For example, if we have:

led: led {
    ...
    pwms = <&pwm0 1 PWM_POLARITY_NORMAL>;
    ...
};

One can do now in the application:

struct pwm_dt_spec led = PWM_DT_SPEC_GET(DT_NODELABEL(led));

pwm_set_usec_dt(&led, LED_PERIOD_USEC, LED_PULSE_USEC);

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: basic: fade_led: use pwm_dt_spec

Simplify sample by using pwm_dt_spec.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: basic: rgb_led: use pwm_dt_spec

Simplify the sample by using pwm_dt_spec.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: basic: servo_motor: use pwm_dt_spec

The sample did not use PWM in a proper way: it relied on a PWM device
handle, but channel or flags were hardcoded in the code. This patch
changes the sample to use a local binding ("pwm-servo") where the `pwms`
property is defined. This allows to make use of pwm_dt_spec facilities,
reducing the overall sample complexity and making it more portable
without editing the source code. The custom binding also requires to
provide the minimum/maximum pulse width so that different servos can be
easily plugged in without the need to edit sources.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: led: led_pwm: use pwm_dt_spec

Simplify the driver by using pwm_dt_spec.

TODO: decide if pwm_dt_spec should also store period.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic reviewed (commented):
Looks quite reasonable to me; +1 on the approach

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -221,17 +221,17 @@ static int get_tim_clk(const struct stm32_pclken *pclken, uint32_t *tim_clk)
 
-	if (pwm < 1u || pwm > TIMER_MAX_CH) {
-		LOG_ERR("Invalid channel (%d)", pwm);
+	if (channel < 1u || channel > TIMER_MAX_CH) {
```
Brings a conflict with existing local variable "channel"

gmarull replied:
fixed, thanks

henrikbrixandersen commented:
I like it.

mbolivar-nordic commented on samples/basic/servo_motor/boards/bbc_microbit.overlay:
```diff
@@ -1,7 +1,8 @@
 / {
-	aliases {
-		pwm-servo = &sw_pwm;
+	servo: servo {
```
This looks like a use case for `/zephyr,user` rather than a custom node like this.

```devicetree
/ {
        zephyr,user { pwms = < ... >; };
};
```

Then you don't need a binding.

gmarull replied:
I prefer to introduce a binding (added now min/max pulse width there as well). Ideally the sample should have a small driver with a custom servo API, this is how things would probably be done in a real application. What sample does is okay for a quick demo, though.

carlescufi commented:
API meeting:

> * Decide on API naming. Drop `_pin`? Use `_channel`..?

Drop the `_pin_` infix in the function names and rename the parameter `pwm` to `channel`.

> * Should `pwm_dt_spec` include the period? Some PWM specifiers (e.g. STM32) have the period cell, some others don't.

Yes, to align with Linux and to make it portable between boards.

anangl commented on drivers/pwm/pwm_led_esp32.c:
```diff
@@ -374,7 +373,7 @@ static int pwm_led_esp32_get_cycles_per_sec(const struct device *dev,
 
 static const struct pwm_driver_api pwm_led_esp32_api = {
-	.pin_set = pwm_led_esp32_pin_set_cycles,
+	.set_cycles = pwm_led_esp32_set_cycles_cycles,
```
```suggestion
	.set_cycles = pwm_led_esp32_set_cycles,
```

gmarull replied:
fixed

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -163,8 +163,8 @@ static int pwm_nrfx_pin_set(const struct device *dev, uint32_t pwm,
 	 * nrfx_pwm_is_stopped may cause race condition if the pwm_nrfx_pin_set
```
pwm_nrfx_pin_set -> pwm_nrfx_set_cycles

gmarull replied:
adjusted

anangl commented on samples/basic/servo_motor/README.rst:
```diff
@@ -19,29 +19,22 @@ to modify the source code if you are using a different servomotor.
+       servo: servo {
+           compatible = "pwm-servo";
+           pwms = <&pwm0 1 20000000 PWM_POLARITY_NORMAL>;
+       };
```
The `min-pulse-usec` and `max-pulse-usec` properties are required now, so they should be added also here.

gmarull replied:
added

anangl commented on samples/basic/servo_motor/dts/bindings/pwm-servo.yaml:
```diff
@@ -0,0 +1,30 @@
+  PWM-driven servo motor. Example definition:
+
+  servo {
+    compatible = "servo";
```
Incorrect compatible string.
But do we actually need an example here? We've got one in README.

gmarull replied:
removed example as there's one on the README

anangl commented on include/zephyr/drivers/pwm.h:
```diff
@@ -263,15 +455,36 @@ static inline int pwm_pin_set_usec(const struct device *dev, uint32_t pwm,
+ *
+ * @return A value from pwm_set_usec().
+ */
+static inline int pwm_set_usec_dt(const struct pwm_dt_spec *spec,
+				  uint32_t pulse)
```
I was wondering if `*_dt` functions that allow setting also `period` could be provided, too (for example, we could have `pwm_set_usec_pulse_dt` for setting pulse only and `pwm_set_usec_dt` for setting both). Those could be useful in cases   like blinky_pwm or when PWM drives a buzzer.

gmarull replied:
updated PR, please review

gmarull commented on samples/basic/servo_motor/README.rst:
```diff
@@ -19,29 +19,22 @@ to modify the source code if you are using a different servomotor.
-   	};
+       servo: servo {
+           compatible = "pwm-servo";
+           pwms = <&pwm0 1 20000000 PWM_POLARITY_NORMAL>;
```
adjusted, forgot to update

gmarull committed:
samples: basic: blinky_pwm: use pwm_dt_spec

Simplify the sample by using pwm_dt_spec facilities.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: release-notes: inform anout PWM changes

- Inform about deprecated functions
- Mention the pwm_dt_spec additions

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: always use nanoseconds for set

In order to be consistent with what is possible in Devicetree, always
take a period in nanoseconds. Other scales or units may be specified by
using, e.g., the PWM_MSEC() macros (all of them converting down to
nanoseconds). This change then deletes the "_nsec" and "_usec" versions
of the pwm_set call.

Note that this change limits the period to UINT32_MAX nanoseconds,
~4.3s. PWM is, in generali, used with periods below the second so it
should not be a problem.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

anangl commented on samples/bluetooth/mesh_demo/src/microbit.c:
```diff
@@ -129,13 +129,14 @@ void board_play_tune(const char *str)
 		if (period) {
-			pwm_set_usec(pwm, BUZZER_PWM_CHANNEL, period, period / 2U, 0);
+			pwm_set(pwm, BUZZER_PWM_CHANNEL, PWM_USEC(period),
+				PWM_USEC(period / 2U), 0);
```
```suggestion
				PWM_USEC(period) / 2U, 0);
```
to get e.g. 1702500 instead of 1702000 when `period` is 3405. No one will actually hear the difference in these samples, but in case their code was to be taken as a pattern, it would be better to avoid integer truncation.

gmarull replied:
adjusted, thanks

gmarull commented on samples/basic/blinky_pwm/README.rst:
```diff
@@ -29,11 +29,7 @@ controlling this LED must be configured using the ``pwm_led0`` :ref:`devicetree
-.. code-block:: none
-
-   Unsupported board: pwm_led0 devicetree alias is not defined
+board.
```
removed

kt-peters commented on drivers/pwm/pwm_shell.c:
```diff
@@ -23,7 +23,7 @@ struct args_index {
 static const struct args_index args_indx = {
 	.device = 1,
-	.pwm = 2,
+	.channel = 2,
```
`struct args_index` has no field `.channel`. This file can no longer be compiled.

gmarull replied:
thanks for spotting this issue, fix https://github.com/zephyrproject-rtos/zephyr/pull/45202
